Python/eof.py

```diff
@@ -0,0 +1,198 @@
+            of random spatially sampled design
+            value grid cells
+        index (numpy.ndarray): index locations 
+            of y_sample from original ensemble size
+    """
+    Y_ens = ensemble_reshape(data_cube)
+    land_idx = get_land(Y_ens)
+    # select only land
+    Y_land = Y_ens[~land_idx.any(axis=1)]
+
+    p, n = Y_land.shape
+
+    # randomly select an ensemble member
+    # to spatially sample
+    i = np.random.randint(0, n-1)
+    y_land = Y_land[:, i]
+
+    # select random indeces
+    index = rand_sample_index(p, frac)
+    y_sample = y_land[index]
+
+    return y_sample, index
+
+def ensemble_to_eof(ensemble_arr):
+    """Perform EOF/PCA dimensionality reduction
+    on ensemble array 
+    --------------------------------
+    Args:  
+        ensemble_arr (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    Returns:
+        eofs (numpy.ndarray): EOF transformed
+            ensemble_arr sorted by max variance 
+            explained
+    """
+    
+    # select only land from ensemble
+    land_idx = get_land(ensemble_arr)
+    ensemble_arr = ensemble_arr[~land_idx.any(axis=1)]
+    
+    skpca = pca.PCA()
+    
+    # transform to get proper shape for EOF dim reduction
+    skpca.fit_transform(ensemble_arr.T)
+    
+    # transform back to familiar shape n x p
+    eofs = skpca.components_.T
+    
+    return eofs
+
+def regress_eof(eofs, obs):
+    """Perform a linear regression between
+    EOFs and observations 
+    --------------------------------
+    Args:  
+        eofs (numpy.ndarray): EOF transformed
+            ensemble_arr containing same grid cells 
+            sample as observations 
+        obs (numpy.ndarray): Gridded observations
+            either from pseudo or real observations
+    Returns:
+        model (scikit-learn model obj): fitted 
+            model object to eofs and obs
+    """
+    lm = linear_model.LinearRegression()
+    model = lm.fit(eofs, obs)
+    print("Model score:", model.score(eofs, obs))
+    return model
+
+def predict_dv(model, eofs_of_model):
+    """Reconstruct design value field from
+    full EOF sample 
+    --------------------------------
+    Args:  
+        model (scikit-learn model obj): fitted 
+            model object to eofs and obs
+        eofs_of_model (numpy.ndarray): EOF transformed
+            ensemble_arr containing all grid
+            cells in design value field
+    Returns:
+        (numpy.ndarray): predicted design
+            value field
+    """
+    return model.predict(eofs_of_model)
+
+def reconstruct_eof_full(predictions, data_cube):
+    """Add EOF reconstructed predictions
+    to all grid cells of CanRCM4 model and 
+    add to data cube 
+    --------------------------------
+    Args:  
+        predictions (numpy.ndarray): predicted design
+            value field
+        data_cube (xarray Dataset): datacube 
+            containing an ensemble of CanRCM4 models
+    Returns:
```

Comment:
There's a big red flag here that you should consider either changing or putting a big warning on the function and that is the use of ***side effects***. Ideally, a function should accept parameters and return a result. *Modifying* one (or more) of the parameters (as you do with `data_cube`) is usually not good practice (aside from some limited specific cases in low-level, high-performance code).

Could you simply return `full_arr[:, 0]` from this function? And if it's important to have it as part of the data cube, do the appending as a data management action at the caller level?